### PR TITLE
[11.x] Handle circular references in model serialization

### PR DESCRIPTION
**Pull request by samlev**
_This is a companion PR to #51582 which resolves the issues raised around serialization of models that have circular references in relations._

# TL;DR:
This PR makes some minor changes to how the main recursive operations on Models work to prevent getting into infintite recursion down a stack of circular references, which leads to a stack overflow. It does this by preventing recursive calls to the same method on an object.

# What are circular references?
Put simply - circular references occur when two objects hold a reference to each other. In the context of this PR, it's when a model holds a relationship to another model which in turn holds a relationship back to the first _instance_. For example, if you had a user with multiple posts, and you wanted to give each post a reference back to the original user, you might do something like this:
```php
$user->posts->each->setRelation('user', $user);
```
This would mean that each post has a link back to the original user, not to another _copy_ of the user.

## Why would you want to do this?
There are many reasons, but the most common reason that I encounter is authorization. What if you had a policy similar to this for posts:
```php
class PostPolicy
{
    // ...
    public function view(User $user, Post $post): bool
    {
        return $post->published_at?->greaterThan(now()) || $post->user->is($user);
    }
    //...
}
```
Note that the post references it's own copy of the user object. When you're looping through posts on an index page you can easily eager-load the user (`Post::with(['user'])->get()`), but if you're pulling the posts for a _particular_ user, you either have to:

* Eager-load the user on the posts (`$user->load(['posts.user']);`) which leaves you with multiple copies of the user that you already had.
* Deal with N+1 queries as each post runs a query to pull out its user - again, something that you had already pulled out of the database.
* Set the relation to the existing model, and end up with a circular reference.

Setting the relation is the most memory efficient, database efficient, and has some other benefits like being able to naviagate back and forth between the user and posts without generating more queries, e.g.
```html
<!-- user/post-index.blade.php -->
@foreach($user->posts as $idx => $post)
  <x-post :$post :position="$idx + 1" />
@endforeach

<!-- components/post.blade.php -->
<a href="{{ route('user.posts.edit', [$post->user, $post]) }}">{{ $post->title }}</a>
({{ $position }} / {{ $post->user->posts->count() }})
```
## How does this cause problems?
For the most part, it doesn't. I've used this pattern many times on many projects, and PR #51582 would make it possible to set the inverse relation automatically in a number of situations.

The issue is when it comes to serializing a model that has circular relations. When you call `toJson()` or `toArray()` on a model with circular relations, or you attempt to send that model to a queue, or you simply want to just call `push()` to save the model and any of it's dependent children. Essentially any time where it tries to walk down the stack of relations recursively, you'll end up in this loop:
```php
>>> dump($user->toArray());
[
    'id' => 1,
    'posts' => [
        0 => [
            'id' => 1234,
            'user' => [
                'id' => 1,
                'posts' => [
                    0 => [
                        'id' => 1234,
                        'user' => [
                            'id' => 1,
                            'posts' => [// ... all the way to a stack overflow
```
I don't know how many people have experienced a SegFault in PHP, but it's... not a fun issue to diagnose.

## How does this PR fix that issue?
This adds a new internal method to models called ~~`once()`~~ `withoutRecursion()`. This keeps track of the call stack and will prevent the same method from being called on the same _instance_ of an object more than once _within that stack_. So instead of that call stack above, you end up with something like this:
```php
>>> dump($user->toArray());
[
    'id' => 1,
    'posts' => [
        0 => [
            'id' => 1234,
            'user' => [
                'id' => 1,
            ],
        ],
        1 => [
            'id' => 2345,
            'user' => [
                'id' => 1,
            ],
        ],
        2 => [
            'id' => 3456,
            'user' => [
                'id' => 1,
            ],
        ],
    ]
];
```
Because `toArray()` was called on `$user` at the top level, it doesn't get called again, so it doesn't dig down again, but because each _post_ is a unique object, `toArray()` can get called on each of them.

# A Couple Of Questions That You Might Want Answered
### Is this backwards compatible?
Yes. This won't break any existing code because any code that would have encoutnered this problem wouldn't be working. ~~The only concerns are if someone has already defined a `once()` method on a model (in which case I'm happy to change this method name if you'd prefer - I was just keeping consistency with the other existing functionality for the `once()` helper)~~.

### Why not use the `once()` helper?
For two major reasons:

1. **It's not recursion safe**

It expects a call to complete so that it can store the value before it can prevent subsequent calls to the method. This means that if the method is recursive, it will never complete. The change that I've made here accepts a "default" value which will be stored as the return value _before_ the method is called, meaning that any recursion will return early with the default value instead of continuing to dig itself deeper.

2. **`once()` would keep the result for the rest of the request, even if the model changed**

We don't want to prevent `->toArray()` from changing its result if the models or relations change. Each time it gets called we want to actually pull the result at that point in time. We only want the "once" behaviour to persist _within the context of that one call stack_. This is particularly important for `->push()` - you don't want to store the result of `$model->push()` and never actually call it again.

### Has anyone other than you even encoutnered this?
Yep! @reinink has talked about [Optimizing circular relationships in Laravel](https://reinink.ca/articles/optimizing-circular-relationships-in-laravel), @stancl made a package to add [`hasManyWithInverse()`](https://github.com/archtechx/laravel-hasmanywithinverse), and @stayallive extended that, with an specific note about [dealing with recursion](https://github.com/stayallive/laravel-inverse-relations?tab=readme-ov-file#caveats).

Even if the `inverse()` pull request never gets merged, I believe that this would still be a valuable fix for models.

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Concerns/PreventsCircularRecursion.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    /**
+     * Prevent a method from being called multiple times on the same object within the same call stack.
+     *
+     * @param  callable  $callback
+     * @param  mixed  $default
+     * @return mixed
+     */
+    protected function once($callback, $default = null)
```
Comment (Tofandel) [inline src/Illuminate/Database/Eloquent/Concerns/PreventsCircularRecursion.php:47]: I do think following laravel's pattern this should be renamed to `withoutInfiniteRecursion` or the like

As you've explained it doesn't really have much to do with the other once helper and could be confusing

Comment (samlev) [reply src/Illuminate/Database/Eloquent/Concerns/PreventsCircularRecursion.php:47]: I had actuall named it `withoutRecursion()` originally then changed it to `once()`

File: src/Illuminate/Database/Eloquent/Model.php [hunk 2]
```diff
@@ -1644,7 +1647,10 @@ public function callNamedScope($scope, array $parameters = [])
      */
     public function toArray()
     {
-        return array_merge($this->attributesToArray(), $this->relationsToArray());
+        return $this->once(
+            fn () => array_merge($this->attributesToArray(), $this->relationsToArray()),
+            $this->attributesToArray(),
+        );
```
Comment (Tofandel) [inline src/Illuminate/Database/Eloquent/Model.php:1653]:
```suggestion
        $attributes = $this->attributesToArray();
        return $this->once(
            fn () => array_merge($attributes, $this->relationsToArray()),
            $attributes,
        );
```

This kind of methods are a bit expensive so I would avoid calling it twice

File: src/Illuminate/Database/Eloquent/Concerns/PreventsCircularRecursion.php (hunk 1 above)
Comment (Tofandel) [reply src/Illuminate/Database/Eloquent/Concerns/PreventsCircularRecursion.php:47]: That also works, as long as it makes the code more readable in `Model` so you can understand without having to take a look inside the function

File: src/Illuminate/Database/Eloquent/Model.php (hunk 2 above)
Comment (samlev) [reply src/Illuminate/Database/Eloquent/Model.php:1653]: It actually winds up only being called twice for the entire stack, no matter how many times it gets called (because the second call is what gets cached).

File: src/Illuminate/Database/Eloquent/Concerns/PreventsCircularRecursion.php (hunk 1 above)
Comment (samlev) [reply src/Illuminate/Database/Eloquent/Concerns/PreventsCircularRecursion.php]: I've updated this back to `withoutRecursion()`

File: src/Illuminate/Database/Eloquent/Model.php (hunk 2 above)
Comment (samlev) [reply src/Illuminate/Database/Eloquent/Model.php]: I've updated the default to accept a callback, and cache that value.

Your suggested change wouldn't prevent `attributesToArray()` from being called on each recursive call to `toArray()` - it would just ignore it on subsequent calls.

**Comment by Jubeki** [comment]
I like this PR in general!

Are there performance / memory implications with these changes? Maybe some benchmarks would be great with the before and after comparison.

**Comment by Jubeki** [comment]
> That's a very difficult question to answer because the "before" case that this PR is targeting would have resulted in a stack overflow, so the test would be "it works at all" vs "Segmentation Fault".

Yeah, I meant regarding a non-circular `toArray` call, because it also uses the without Recursion part.

**Comment by samlev** [comment]
> Yeah, I meant regarding a non-circular `toArray` call, because it also uses the without Recursion part.

So there would be extra memory usage due to `attributesToArray()` being cached _in case_ there's recursion. I could mitigate against it but it would only add extra complexity to address something that is probably not (and I'm already not _100% happy_ with the current increase to complexity for `toArray()`, but I'm hedging memory usage against processing time in response to @Tofandel 's review).

**Comment by taylorotwell** [comment]
@samlev really dig the PR, but unfortunately it is a pretty big hit on performance. For example, just seeding 1000 users into the database and doing `return User::all()->toArray()` from a route. That `toArray` call is 2x slower on this PR compared to current `11.x` branch. I jump from roughly 40-50ms on my machine to about 100ms on this PR.

**Comment by samlev** [comment]
@taylorotwell There's definitely some ways to clean it up more. I'm guessing that the speed issue is the additional call to `attributesToArray()` which could actually be done as an "on demand"/"when needed" - I was just trying to avoid adding even more complexity.

**Comment by taylorotwell** [comment]
@samlev we may just have to give the extra complexity a shot - hopefully won't be too bad. I don't think we can stomach a 2x performance hit for all Eloquent serialization.

**Comment by samlev** [comment]
@taylorotwell this new update _should_ negate the performance penalty for non-circular relations (and should more or less resolve @Jubeki 's concerns about increased memory usage).

The only other "improvement" that I can think of would be to get extremely meta by _forcing_ `toArray()` to be called recursively:
```php
public function toArray()
{
    return $this->withoutRecursion(
        fn () => array_merge($this->toArray(), $this->relationsToArray()),
        fn () => $this->attributesToArray(),
    );
}
```
which would have the effect of only calling `attributesToArray()` _once_ to fulfill the second `toArray()` call and give the default value. So the call stack would go something like this (visualising recursive call stacks in text is hard...):
```php
toArray() {
  withoutRecursion() {
    toArray() {
      withoutRecursion() {
        attributesToArray();
      }
    }
    relationsToArray();
  }
}
```

**Comment by taylorotwell** [comment]
@samlev Indeed I don't see the performance regression anymore. Nice! Will do a final review in the morning and try to get this merged. 👍

**Comment by rodrigopedra** [comment]
Good to see this being handled, but isn't using a flag a better approach? Regarding performance?

Something like this:

```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Item extends Model
{
    // recursion flag
    protected $visited = false;

    public function item(): BelongsTo
    {
        return $this->belongsTo(Item::class);
    }

    public function items(): HasMany
    {
        return $this->hasMany(Item::class);
    }

    public function toArray(): array
    {
        if ($this->visited) {
            // TODO: improve the fallback case, maybe an array 
            // with a single marker value to filter out later?
            return [];
        }

        $this->visited = true;

        try {
            return parent::toArray();
        } finally {
            $this->visited = false;
        }
    }
}
```

With this test code:

```php
<?php

use App\Models\Item;
use Illuminate\Support\Facades\Artisan;

Artisan::command('circular', function () {
    $parent = Item::query()->create();
    $parent->items()->create();
    $parent->items()->create();
    $parent->items()->create();

    $parent->load(['items']);

    // set parent instance manually to create the circular reference
    $parent->items->each->setRelation('item', $parent);

    $this->info($parent->toJson(\JSON_PRETTY_PRINT));

    // print twice to show the visited property is cleared nicely
    $this->info($parent->toJson(\JSON_PRETTY_PRINT));
});
```

Yields this output:

```bash
$ php artisan circular
{
    "updated_at": "2024-08-22T05:28:24.000000Z",
    "created_at": "2024-08-22T05:28:24.000000Z",
    "id": 1,
    "items": [
        {
            "id": 2,
            "item_id": 1,
            "created_at": "2024-08-22T05:28:24.000000Z",
            "updated_at": "2024-08-22T05:28:24.000000Z",
            "item": []
        },
        {
            "id": 3,
            "item_id": 1,
            "created_at": "2024-08-22T05:28:24.000000Z",
            "updated_at": "2024-08-22T05:28:24.000000Z",
            "item": []
        },
        {
            "id": 4,
            "item_id": 1,
            "created_at": "2024-08-22T05:28:24.000000Z",
            "updated_at": "2024-08-22T05:28:24.000000Z",
            "item": []
        }
    ]
}
{
    "updated_at": "2024-08-22T05:28:24.000000Z",
    "created_at": "2024-08-22T05:28:24.000000Z",
    "id": 1,
    "items": [
        {
            "id": 2,
            "item_id": 1,
            "created_at": "2024-08-22T05:28:24.000000Z",
            "updated_at": "2024-08-22T05:28:24.000000Z",
            "item": []
        },
        {
            "id": 3,
            "item_id": 1,
            "created_at": "2024-08-22T05:28:24.000000Z",
            "updated_at": "2024-08-22T05:28:24.000000Z",
            "item": []
        },
        {
            "id": 4,
            "item_id": 1,
            "created_at": "2024-08-22T05:28:24.000000Z",
            "updated_at": "2024-08-22T05:28:24.000000Z",
            "item": []
        }
    ]
}
```

Of course, there is this dangling array on the child elements' parent references, but this is just a proof of concept.

I thought on hacking with the `Model@relationsToArray()` method to filter them out, but didn't want to invest time on it before asking for feedback.

**Comment by rodrigopedra** [comment]
Yes, adding a marker will remove the dangling ones:

```php
<?php

namespace App;

enum Visited
{
    case VISITED;
}
```

Changed model to account for the marker:

```php
<?php

namespace App\Models;

use App\Visited;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Item extends Model
{
    // recursion flag
    protected $visited = false;

    public function item(): BelongsTo
    {
        return $this->belongsTo(Item::class);
    }

    public function items(): HasMany
    {
        return $this->hasMany(Item::class);
    }

    public function toArray(): array
    {
        if ($this->visited) {
            return [Visited::VISITED];
        }

        $this->visited = true;

        try {
            return parent::toArray();
        } finally {
            $this->visited = false;
        }
    }

    public function relationsToArray(): array
    {
        // filters out visited relations
        return \array_filter(parent::relationsToArray(), fn ($item) => $item !== [Visited::VISITED]);
    }
}
```

Updated output (with a single `toJson()` call):

```bash
$ php artisan circular
{
    "updated_at": "2024-08-22T05:43:34.000000Z",
    "created_at": "2024-08-22T05:43:34.000000Z",
    "id": 1,
    "items": [
        {
            "id": 2,
            "item_id": 1,
            "created_at": "2024-08-22T05:43:34.000000Z",
            "updated_at": "2024-08-22T05:43:34.000000Z"
        },
        {
            "id": 3,
            "item_id": 1,
            "created_at": "2024-08-22T05:43:34.000000Z",
            "updated_at": "2024-08-22T05:43:34.000000Z"
        },
        {
            "id": 4,
            "item_id": 1,
            "created_at": "2024-08-22T05:43:35.000000Z",
            "updated_at": "2024-08-22T05:43:35.000000Z"
        }
    ]
}
```

Of course, instead of overriding those methods on user land, we can consider incorporating them upstream.

**Comment by rodrigopedra** [comment]
Just for easiness to reproduce (e.g. copy and paste), here is the migration I used:

```php
<?php

use App\Models\Item;
use Illuminate\Database\Migrations\Migration;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;

return new class extends Migration
{
    public function up(): void
    {
        Schema::create('items', function (Blueprint $table) {
            $table->id();
            $table->foreignIdFor(Item::class)->nullable()->constrained();
            $table->timestamps();
        });
    }

    public function down(): void
    {
        Schema::dropIfExists('items');
    }
};
```

**Comment by rodrigopedra** [comment]
Reproducible repo: https://github.com/rodrigopedra/circular

P.S.: moved overrides to a trait

**Comment by samlev** [comment]
@rodrigopedra the minor difference between the two implementations here is that the nested children will still have the relation, they just won't have the relation of the relations.

So with your example (timestamps removed for brevity):
```json
{
    "id": 1,
    "items": [
        {
            "id": 2,
            "item_id": 1
        },
        {
            "id": 3,
            "item_id": 1
        },
        {
            "id": 4,
            "item_id": 1
        }
    ]
}
```
The implemntation in this PR would still have the "item" on each of the children:
```json
{
    "id": 1,
    "items": [
        {
            "id": 2,
            "item_id": 1,
            "item": {
                "id": 1
            }
        },
        {
            "id": 3,
            "item_id": 1,
            "item": {
                "id": 1
            }
        },
        {
            "id": 4,
            "item_id": 1,
            "item": {
                "id": 1
            }
        }
    ]
}
```
This doesn't seem _super_ important, but it's more "accurate" as to the actual state of each object, and more consistent with the result if you had eager-loaded the relations (e.g. `Item::with('items.item')`).

There's more to it, but we also have the potential for a method that causes recursion to call another method that might cause recursion, and we don't want to prevent that second call from happening but we still want to be sure that everything is cleaned up after any individual call to `toArray()` or `push()` or `getQueueableRelations()`.

Doing it with a static weak map also opens up possibilities with inspecting levels of recursion (e.g. `static::getRecursionCache()->count()` would tell you how many _individual objects_ of the current class are currently in play in the current call stack - something that might be useful for diagnostics or observability, or even just limiting the size of the call stack). WeakMaps "clean up" after themselves - or more to the point they don't prevent objects from being garbage collected - so we also don't really have a risk of orphaned records, or holding on to extra memory for any longer than we need to.

**Comment by donnysim** [comment]
@rodrigopedra yours has an edge case when the relation collection is shared while current PR does not:
```php
$sharedItem = new Item(['name' => 'shared']);

$itemA = new Item(['name' => 'a']);
$itemA->setRelation('shared', $sharedItem);
$itemB = new Item(['name' => 'b']);
$itemB->setRelation('shared', $sharedItem);

$items = Collection::make([$itemA, $itemB]);

$parent = new Item(['name' => 'parent']);
$parent->setRelation('items', $items);

$sharedItem->setRelation('items', $items);

dd($parent->toArray());
```
Output:
```
array:2 [
  "name" => "parent"
  "items" => array:2 [
    0 => array:2 [
      "name" => "a"
      "shared" => array:2 [
        "name" => "shared"
        "items" => array:2 [
          0 => array:1 [
            0 => App\Visited {#1671
              +name: "VISITED"
            }
          ]
          1 => array:1 [
            "name" => "b"
          ]
        ]
      ]
    ]
    1 => array:2 [
      "name" => "b"
      "shared" => array:2 [
        "name" => "shared"
        "items" => array:2 [
          0 => array:1 [
            "name" => "a"
          ]
          1 => array:1 [
            0 => App\Visited {#1671}
          ]
        ]
      ]
    ]
  ]
]
```

File: tests/Database/DatabaseConcernsPreventsCircularRecursionTest.php [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    public function testRecursiveCallsArePreventedWithoutPreventingSubsequentCalls()
+    {
+        $instance = new PreventsCircularRecursionWithRecursiveMethod();
+
+        $this->assertEquals(0, PreventsCircularRecursionWithRecursiveMethod::$globalStack);
+        $this->assertEquals(0, $instance->instanceStack);
+
+        $this->assertEquals(0, $instance->callStack());
```
Comment (rodrigopedra) [inline tests/Database/DatabaseConcernsPreventsCircularRecursionTest.php:24]: Why is this zero?

On the first iteration, the callback is executed and `$this->instanceStack` becomes `1` and returns the recursive call.

On the recursive call, the recursion is guarded, then it returns `$this->instanceStack`, which should be `1`, right?

**Comment by rodrigopedra** [comment]
@samlev thanks, I noticed that when looking at the tests. And you are right, that'd be more accurate.

@donnysim while digging into the PR's code, I moved away from using this marker. But, thank you so much for reviewing it, and for the insights.

@samlev I updated my test repo, and also created a fork to use the same test set as yours (Minus the test cases that, I commented out that I don't quite get the assertion being `0`):

https://github.com/rodrigopedra/framework/commit/e2616f3edb38080768ba4350d3cffc5613424f21

I didn't want to push a new PR, as I don't want to overtake this one.

I understood from your test cases that my alternative implementation was missing the mixed function recursion. So I incorporated the stack trace you have in yours.

Can you review it? I believe it is a simpler approach, and avoids a static global `WeakMap`.

Nonetheless, thank you very much for this PR, I had to deal with manually unsetting relations on many projects before.

**Comment by rodrigopedra** [comment]
@donnysim, on my POC repository, I added a second command to account for the use case you mentioned.

Follow its README and run `php artisan shared`

Thanks again for the insights.

File: tests/Database/DatabaseConcernsPreventsCircularRecursionTest.php (hunk 3 above)
Comment (samlev) [reply tests/Database/DatabaseConcernsPreventsCircularRecursionTest.php:24]: The "default" value that gets returned when it recurses is `$this->instanceStack` as it was passed in when it was originally called (i.e. `0`). If the default was a callback, it would resolve later and return `1`.

So in essence, the first call does this:
```php
public function callStack(): int
    return $this->withoutRecursion(
        function () {
            static::$globalStack++;
            $this->instanceStack++;

             return $this->callStack(); // gets the cached/default value that was passed in below
        },
        $this->instanceStack // currently 0
    );
}
```

If the call was made using a callback method that resolve _later_: 
```php
public function callStack(): int
    return $this->withoutRecursion(
        function () {
            static::$globalStack++;
            $this->instanceStack++;

             return $this->callStack(); // calls the callback which gets the current version now (which is 1), then caches that value for subsequent calls
        },
        fn () => $this->instanceStack
    );
}
```

Comment (samlev) [reply tests/Database/DatabaseConcernsPreventsCircularRecursionTest.php:24]: @rodrigopedra thanks for reminding me to add some additional tests for the "default callback" feature

Comment (rodrigopedra) [reply tests/Database/DatabaseConcernsPreventsCircularRecursionTest.php:24]: I get what you mean, it behaves as it is avoiding the recursion altogether, right?

But either way, it still iterates the recursion a second time, but instead of returning the second call's `$default` value, it pulls the cached `$default` value from the first call, right?

I mean:

```php
$instance->callStack(); // first time

// Implementation calls:
$this->withoutRecursion(fn () => ..., 0);

// Inside the callback it calls... 
$instance->callStack(); // ...a second time time
// from the return statement, which is a recursive call

// Implementation calls this, again...
$this->withoutRecursion(fn () => ..., 1);
// ...where it detects a recursive call, ignores the current default
// and pulls the original default from the cache from the first call
```

IMO, I find it a bit odd, from my gut I'd expect the `$default` to be used when the `$callback` can't be used anymore, which would be the `$default` from the second call.

But I see what you mean, it is more like, "when we can't recurse anymore, use this default instead", and then you use the cached one.

I guess it is just different point of views.

Thanks for your patience!

Comment (samlev) [reply tests/Database/DatabaseConcernsPreventsCircularRecursionTest.php:24]: Yeah, there's potentially a different or "better" name for that variable, but nothing that sprang to mind as being particularly "Laravel-ish". Maybe `$onRecursion`, or `$recurseWith`, or `$andThen`, but nothing that really _stands out_.

**Comment by Tofandel** [comment]
The big problem with the flag is if you call the method multiple times after having changed attributes for example, which in theory the `withoutRecursion` doesn't have

It's a bit of a functional programming vs OOP

**Comment by rodrigopedra** [comment]
@Tofandel as an internal/`protected` property, that would only happen if someone overrides it manually in an unintended way. 

Yet, the same can happen with the static cache. One could just assign a new `WeakMap` instance to the cache and blow things up.

When I saw the tests I moved away to a single flag, due to the mixed recursive method call, as a single flag prevented that use case.

(check the `DatabaseConcernsPreventsCircularRecursionTest` and look for the `PreventsCircularRecursionWithRecursiveMethod@callOtherStack` method)

But I'd still go for a per-instance stack than a global `WeakMap`. That is just as a matter of preference, the current state of this PR is a great addition, IMO.

**Comment by Tofandel** [comment]
I was talking about your original implementation in comments, your new implementation is almost the same as the one proposed in this PR and uses a WeakMap of the callstack as well

**Comment by rodrigopedra** [comment]
@Tofandel it has a different approach for not caching the first ever `$default` value.

On this PR's implementation, the first ever `$default` value is cached for future reference, on mine it uses the then current value from the next recursive call.

But nonetheless, apart from the cross recursive calls between different functions, which clearly don't work with the single flag approach, I wonder how keeping a map of call stack hashes is much different than the flag approach, regarding your functional vs OOP comment.

I moved away from the single flag implementation to accommodate the cross recursive calls for different functions. Although I think that for the serialization use case it would be totally fine, as none of the currently affected methods reference each other.

It is ok if you don't want to stretch this further, and thanks again for reviewing my code =)

**Comment by taylorotwell** [comment]
Thanks @samlev - really well done PR.
